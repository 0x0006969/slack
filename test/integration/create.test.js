const request = require('supertest');
const nock = require('nock');

const { probot, models } = require('.');
const fixtures = require('../fixtures');

const {
  SlackUser,
  GitHubUser,
  Subscription,
  Installation,
} = models;

describe('Integration: Creating issues from Slack', () => {
  let workspace;
  let githubUser;
  let slackUser;
  let installation;
  beforeEach(async () => {
    const { SlackWorkspace } = models;
    workspace = await SlackWorkspace.create({
      slackId: 'T0001',
      accessToken: 'xoxa-token',
    });

    githubUser = await GitHubUser.create({
      id: 1,
      accessToken: 'secret',
    });

    slackUser = await SlackUser.create({
      slackId: 'U2147483697', // same as in fixtures.slack.command
      slackWorkspaceId: workspace.id,
      githubId: githubUser.id,
    });

    installation = await Installation.create({
      githubId: 1,
      ownerId: 1337,
    });
  });
  test('works in channel with one subscription', async () => {
    await Subscription.subscribe({
      creatorId: slackUser.id,
      slackWorkspaceId: workspace.id,
      githubId: 54321,
      channelId: 'C2147483705',
      installationId: installation.id,
    });

    nock('https://api.github.com').get('/repositories/54321').reply(200, {
      full_name: 'kubernetes/kubernetes',
    });

    nock('https://slack.com').post('/api/dialog.open', (body) => {
      expect(body).toMatchSnapshot();
      return true;
    }).reply(200, { ok: true });

    const command = fixtures.slack.command({
      text: 'new issue',
    });

    await request(probot.server).post('/slack/command').send(command)
      .expect(200)
      .expect((res) => {
        expect(res.body).toMatchSnapshot();
      });

    nock('https://api.github.com').get('/repositories/54321').reply(200, {
      name: 'kubernetes',
      owner: {
        login: 'kubernetes',
      },
    });

    nock('https://api.github.com').post('/repos/kubernetes/kubernetes/issues').reply(200);

    // User submits dialog to open issue
    await request(probot.server).post('/slack/actions').send({
      payload: JSON.stringify(fixtures.slack.action.dialogSubmissionSingleRepo()),
    })
      .expect(200)
      .expect((res) => {
        expect(res.body).toMatchSnapshot();
      });
  });

  test('works in channel with multiple subscriptions', async () => {
    await Subscription.subscribe({
      creatorId: slackUser.id,
      slackWorkspaceId: workspace.id,
      githubId: 54321,
      channelId: 'C2147483705',
      installationId: installation.id,
    });

    await Subscription.subscribe({
      creatorId: slackUser.id,
      slackWorkspaceId: workspace.id,
      githubId: 654321,
      channelId: 'C2147483705',
      installationId: installation.id,
    });

    nock('https://api.github.com').get('/repositories/54321').reply(200, {
      full_name: 'kubernetes/kubernetes',
    });

    nock('https://api.github.com').get('/repositories/654321').reply(200, {
      full_name: 'atom/atom',
    });

    nock('https://slack.com').post('/api/dialog.open', (body) => {
      expect(body).toMatchSnapshot();
      return true;
    }).reply(200, { ok: true });

    const command = fixtures.slack.command({
      text: 'new issue',
    });

    await request(probot.server).post('/slack/command').send(command)
      .expect(200)
      .expect((res) => {
        expect(res.body).toMatchSnapshot();
      });

    nock('https://api.github.com').get('/repositories/54321').reply(200, {
      name: 'kubernetes',
      owner: {
        login: 'kubernetes',
      },
    });

    nock('https://api.github.com').post('/repos/kubernetes/kubernetes/issues').reply(200);

    // User submits dialog to open issue
    await request(probot.server).post('/slack/actions').send({
      payload: JSON.stringify(fixtures.slack.action.dialogSubmissionRepoSelection()),
    })
      .expect(200)
      .expect((res) => {
        expect(res.body).toMatchSnapshot();
      });
  });
});
