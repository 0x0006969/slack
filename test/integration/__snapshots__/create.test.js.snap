// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration: Creating issues from Slack a channel with no subscriptions will show an error message 1`] = `
Object {
  "text": "This channel is not subscribed to any repositories",
}
`;

exports[`Integration: Creating issues from Slack repos which are not found are excluded from the list of repos to take action on 1`] = `
Object {
  "dialog": "{\\"callback_id\\":\\"create-issue-dialog\\",\\"title\\":\\"Open new issue\\",\\"submit_label\\":\\"Open\\",\\"elements\\":[{\\"type\\":\\"text\\",\\"label\\":\\"Title\\",\\"name\\":\\"title\\",\\"hint\\":\\"The issue will be opened on atom/atom\\"},{\\"type\\":\\"textarea\\",\\"label\\":\\"Write\\",\\"name\\":\\"body\\",\\"placeholder\\":\\"Leave a comment\\",\\"hint\\":\\"GitHub markdown syntax is supported, although you cannot preview it in Slack.\\"}]}",
  "token": "xoxa-token",
  "trigger_id": "13345224609.738474920.8088930838d88f008e0",
}
`;

exports[`Integration: Creating issues from Slack repos which are not found are excluded from the list of repos to take action on 2`] = `""`;

exports[`Integration: Creating issues from Slack works in channel with multiple subscriptions 1`] = `
Object {
  "dialog": "{\\"callback_id\\":\\"create-issue-dialog\\",\\"title\\":\\"Open new issue\\",\\"submit_label\\":\\"Open\\",\\"elements\\":[{\\"label\\":\\"Repository\\",\\"type\\":\\"select\\",\\"name\\":\\"repository\\",\\"placeholder\\":\\"Select the repository on which to open the issue\\",\\"options\\":[{\\"label\\":\\"kubernetes/kubernetes\\"},{\\"label\\":\\"atom/atom\\"}]},{\\"type\\":\\"text\\",\\"label\\":\\"Title\\",\\"name\\":\\"title\\"},{\\"type\\":\\"textarea\\",\\"label\\":\\"Write\\",\\"name\\":\\"body\\",\\"placeholder\\":\\"Leave a comment\\",\\"hint\\":\\"GitHub markdown syntax is supported, although you cannot preview it in Slack.\\"}]}",
  "token": "xoxa-token",
  "trigger_id": "13345224609.738474920.8088930838d88f008e0",
}
`;

exports[`Integration: Creating issues from Slack works in channel with multiple subscriptions 2`] = `""`;

exports[`Integration: Creating issues from Slack works in channel with multiple subscriptions 3`] = `Object {}`;

exports[`Integration: Creating issues from Slack works in channel with one subscription 1`] = `
Object {
  "dialog": "{\\"callback_id\\":\\"create-issue-dialog\\",\\"title\\":\\"Open new issue\\",\\"submit_label\\":\\"Open\\",\\"elements\\":[{\\"type\\":\\"text\\",\\"label\\":\\"Title\\",\\"name\\":\\"title\\",\\"hint\\":\\"The issue will be opened on kubernetes/kubernetes\\"},{\\"type\\":\\"textarea\\",\\"label\\":\\"Write\\",\\"name\\":\\"body\\",\\"placeholder\\":\\"Leave a comment\\",\\"hint\\":\\"GitHub markdown syntax is supported, although you cannot preview it in Slack.\\"}]}",
  "token": "xoxa-token",
  "trigger_id": "13345224609.738474920.8088930838d88f008e0",
}
`;

exports[`Integration: Creating issues from Slack works in channel with one subscription 2`] = `""`;

exports[`Integration: Creating issues from Slack works in channel with one subscription 3`] = `Object {}`;
